src/main/java/com/atlassian/db/replica/api/ThrottledCache.java

```diff
@@ -57,36 +62,106 @@ private ThrottledCacheWithTimeout(Clock clock, Duration timeout) {
         @Override
         public Optional<T> get(Supplier<T> supplier) {
             maybeRefresh(supplier);
-            return Optional.ofNullable(value);
+            return Optional.ofNullable(lockValuePair.get().getValue());
         }
 
         @Override
         public Optional<T> get() {
-            return Optional.ofNullable(value);
+            return Optional.ofNullable(lockValuePair.get().getValue());
         }
 
         private void maybeRefresh(Supplier<T> supplier) {
             final Instant deadline = clock.instant().plus(this.timeout);
-            final boolean hasLock = lock.compareAndSet(null, deadline);
+            final T currentValue = this.lockValuePair.get().getValue();
+            final LockValuePair<T> valueWithLock = new LockValuePair<>(
+                deadline,
+                currentValue
+            );
+            final LockValuePair<T> currentLockValuePair = this.lockValuePair.get();
+            final boolean hasLock = !currentLockValuePair.isLocked() && this.lockValuePair.compareAndSet(
```

Comment:
here you compare the same value but for previous deadline and new one to get the lock.